packages/hasher/src/lockfile.ts

```diff
@@ -0,0 +1,91 @@
+import path from "path";
+import findUp from "find-up";
+import fs from "fs-extra";
+import { parse as parseYarnLock } from "@yarnpkg/lockfile";
+import { readWantedLockfile } from "@pnpm/lockfile-file";
+import { Dependencies } from "./resolveExternalDependencies";
+import { nameAtVersion } from "./helpers";
+
+type LockDependency = {
+  version: string;
+  dependencies?: Dependencies;
+};
+
+export type ParsedLock = {
+  type: "success" | "merge" | "conflict";
+  object: {
+    [key in string]: LockDependency;
+  };
+};
+
+const memoization: { [path: string]: ParsedLock } = {};
+
+export async function parseLockFile(packageRoot: string): Promise<ParsedLock> {
+  const yarnLockPath = await findUp("yarn.lock", { cwd: packageRoot });
+
+  // First, test out whether this works for yarn
+  if (yarnLockPath) {
+    if (memoization[yarnLockPath]) {
+      return memoization[yarnLockPath];
+    }
+
+    const yarnLock = fs.readFileSync(yarnLockPath).toString();
+    const parsed = parseYarnLock(yarnLock);
+
+    memoization[yarnLockPath] = parsed;
+
+    return parsed;
+  }
+
+  // Second, test out whether this works for pnpm
+  // TODO: support rush common\config\rush\pnpm-lock.yaml
+  let pnpmLockPath = await findUp("pnpm-lock.yaml", { cwd: packageRoot });
```

Comment:
FYI: before pnpm 3.0 its lock file is named `shrinkwrap.yaml`, might need to cover that. As of now, rush uses pnpm 2.15.1 out of the box.